library/kani/src/aliasing.rs

```diff
@@ -0,0 +1,313 @@
+    pub(super) fn push<U>(tag: PointerTag, perm: PermissionByte, pointer: *const U) {
+        // Decide whether to initialize the stacks
+        // for location:location+size_of(U).
+        // Offset has already been picked earlier.
+        unsafe {
+            use self::*;
+            if STATE == MonitorState::ON
+                && pointer_object(MONITORED) == pointer_object(pointer)
+                && pointer_offset(MONITORED) <= std::mem::size_of::<U>()
+            {
+                STACK_TAGS[STACK_TOP] = tag;
+                STACK_PERMS[STACK_TOP] = perm;
+                STACK_TOP += 1;
+            }
+        }
+    }
+
+    /// Run a stack check on the monitored byte for the given
+    /// tag and the given access permission.
+    pub(super) fn stack_check(tag: PointerTag, access: AccessBit) {
+        unsafe {
+            use self::*;
+            if STATE == MonitorState::ON {
+                let mut found = false;
+                let mut j = 0;
+                let mut new_top = 0;
+                crate::assert(STACK_TOP < STACK_DEPTH, "Max # of nested borrows (15) exceeded");
+                while j < STACK_DEPTH {
+                    if j < STACK_TOP {
+                        let id = STACK_TAGS[j];
+                        let kind = STACK_PERMS[j];
+                        if Permission::grants(access, kind) && id == tag {
+                            new_top = j + 1;
+                            found = true;
+                        }
+                    }
+                    j += 1;
+                }
+                STACK_TOP = new_top;
+                STACK_VALID = STACK_VALID && found;
+            }
+        }
+    }
+}
+
+/// Push the permissions at the given location
+fn push<U>(tag: PointerTag, perm: PermissionByte, address: *const U) {
+    self::monitors::push(tag, perm, address)
+}
+
+/// Initialize the local stored at reference if initialized is set to false,
+/// and track it using a monitor when using demonic nondeterminism.
+///
+/// Every function call in the source program stack-allocates
+/// the local variables that it uses; references to these
+/// variables are only valid after these variables are initialized (first written).
+/// Therefore this function can be used by supplying an initialized flag
+/// set to true after the first write, a track flag set to the value
+/// of a query to a demonic nondeterminism oracle (when this feature is used)
+/// and a reference to the stack location.
+#[rustc_diagnostic_item = "KaniInitializeLocal"]
+fn initialize_local<U>(pointer: *const U) {
+    unsafe {
+        let tag = NEXT_TAG;
+        TAGS.set(pointer, tag);
+        PERMS.set(pointer, Permission::UNIQUE);
+        NEXT_TAG += 1;
+        monitors::track_local(tag, pointer);
+    }
+}
+
+#[rustc_diagnostic_item = "KaniStackCheckPtr"]
+fn stack_check_ptr<U>(pointer_value: *const *mut U) {
+    unsafe {
+        let tag = TAGS.get(pointer_value);
+        let perm = PERMS.get(pointer_value);
```

Comment:
please move this to be inside the `if` block. This logic adds unnecessary overhead if this is not the pointer currently being monitored.